src/charstream.c

```diff
@@ -1,17 +1,36 @@
-#include "charstream.h"
+#include <assert.h>
 
+#include "charstream.h"
 #include "utils.h"
 
 int charstream_init(struct charstream *cs, FILE *file)
 {
-	UNIMPLEMENTED();
+	assert(file);
+
+	cs->file = file;
+
+	return 0;
 }
 char charstream_read(struct charstream *cs)
 {
-	UNIMPLEMENTED();
+	char chr;
+	if (fread(&chr, sizeof(char), 1, cs->file) == 0) {
+		if (feof(cs->file)) {
+			chr = CHARSTREAM_EOF;
+		} else {
+			chr = CHARSTREAM_ERR;
+		}
+		clearerr(cs->file);
+	}
+
+	return chr;
 }
 
 char charstream_peek(struct charstream *cs)
 {
-	UNIMPLEMENTED();
+	char chr = charstream_read(cs);
+	if (chr != CHARSTREAM_ERR)
+		fseek(cs->file, -1, SEEK_CUR);
```

Comment:
What about `fgetc(); ungetc();` ? Easier to grasp imo

Reply:
Fixed 5e16e8c